server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -141,25 +145,27 @@ private ContextIndexSearcher(
         MutableQueryTimeout cancellable,
         int minimumDocsPerSlice,
         boolean wrapWithExitableDirectoryReader,
-        ThreadPoolExecutor executor
+        BoundedExecutor executor,
+        boolean forceSequentialCollection
     ) throws IOException {
         // we need to pass the executor up so it can potentially be used as a sliceExecutor by knn search
         super(wrapWithExitableDirectoryReader ? new ExitableDirectoryReader((DirectoryReader) reader, cancellable) : reader, executor);
         setSimilarity(similarity);
         setQueryCache(queryCache);
         setQueryCachingPolicy(queryCachingPolicy);
         this.cancellable = cancellable;
-        this.queueSizeBasedExecutor = executor != null ? new QueueSizeBasedExecutor(executor) : null;
         this.minimumDocsPerSlice = minimumDocsPerSlice;
-        if (executor != null) {
-            this.leafSlices = computeSlices(
-                getLeafContexts(),
-                queueSizeBasedExecutor.threadPoolExecutor.getMaximumPoolSize(),
-                minimumDocsPerSlice
-            );
-            assert (this.leafSlices.length <= executor.getMaximumPoolSize());
-        } else {
+        if (executor == null) {
             this.leafSlices = null;
+        } else {
+            if (forceSequentialCollection) {
+                // we offload to the executor unconditionally, including requests that don't support concurrency
+                // TODO are we good offloading aggs that don't support concurrency to the concurrent executor or is postCollect an issue?
+                this.leafSlices = new LeafSlice[]{new LeafSlice(new ArrayList<>(getLeafContexts()))};
+            } else {
+                this.leafSlices = computeSlices(getLeafContexts(), executor.getBound(), minimumDocsPerSlice);
+                assert (this.leafSlices.length <= executor.getBound());
```

Comment:
could you add an assertion message here? I always find those helpful when I get an angry assertion stacktrace during testing :)